v2/pkg/genruntime/conditions/conditions.go

```diff
@@ -207,17 +207,50 @@ func (c Condition) String() string {
 }
 
 // SetCondition sets the provided Condition on the Conditioner. The condition is only
-// set if the new condition is in a different state than the existing condition of
-// the same type.
+// set if the new condition is different from the existing condition of the same type.
+// See Condition.IsEquivalent and Condition.ShouldOverwrite for more details.
 func SetCondition(o Conditioner, new Condition) {
+	setCondition(o, new, func(new Condition, old Condition) bool { return new.ShouldOverwrite(old) })
+}
+
+// SetConditionReasonAware sets the provided Condition on the Conditioner. This is similar to SetCondition
+// with one difference: SetConditionReasonAware understands common Reasons used by ASO and allows some of them to
+// modify the standard Condition priority rules. This is primarily used to allow the Reconciling condition to overwrite
+// Warning conditions raised by the operator that have been fixed. This is useful because sometimes getting a success or
+// error from Azure can take a long time, and workflows like: submit -> warning -> fix warning -> call Azure -> wait -> success
+// otherwise would continue reporting the Warning Condition until the final success step (possibly many minutes after the
+// warning was resolved).
+func SetConditionReasonAware(o Conditioner, new Condition) {
+	shouldOverwrite := func(new Condition, old Condition) bool {
+		reasonPriority := map[string]int{
+			ReasonReferenceNotFound: -2,
+			ReasonSecretNotFound:    -2,
+			ReasonWaitingForOwner:   -2,
+			ReasonReconciling:       -1,
```

Comment:
Maybe a comment to explain why negative numbers are used. Is it so that zero takes priority over these?

Reply:
And maybe stash the map in a global so we don't create it every time you call `shouldOverwrite()`?

Reply:
Did both of these